### ISSUE TITLE
feat: 基于全唐诗实现 Char-RNN 字符级文本生成

### ISSUE THREAD

**Issue by NayukiChiba** [NayukiChiba/Char-RNN#1]
## 一、任务内容

用 PyTorch 从零实现一个字符级 RNN，以全唐诗为训练数据，让模型学习汉字序列的概率分布，最终能够根据给定的起始字自动续写"诗句"。

```
原始唐诗文本 → 字符编码 → 序列切分 → RNN 训练 → 逐字采样生成
```

## 二、项目架构

```
Char-RNN/
├── main.py                          # 唯一主入口
├── config/
│   ├── __init__.py
│   ├── defaults.py                  # 默认超参
│   └── paths.py                     # 路径常量
├── data/
│   ├── __init__.py
│   ├── charVocab.py                 # 字符 ↔ 索引 双向映射
│   └── dataLoader.py                # 序列切分 + DataLoader 封装
├── models/
├── training/
├── evaluation/
├── inference/
├── cli/
│   ├── __init__.py
│   ├── parse.py                  # 指令式入口（argparse 子命令）
│   └── menu.py                      # 交互式菜单入口
├── notebooks/
│   └── exploration.ipynb            # 数据探索、效果演示
├── tests/
│   └── testModel.py                 # 前向传播 shape 验证
├── outputs/
│   ├── checkpoints/                 # 模型权重
│   ├── logs/                        # 训练日志
│   └── visualizations/             # loss 曲线等图表
├── datasets/                        # 原始数据（.gitignore）
└── pyproject.toml
```

### 模块职责

| 模块 | 职责 |
|------|------|
| `main.py` | 唯一主入口，路由到 CLI 指令模式或交互菜单模式 |
| `config/` | 超参默认值 + 项目路径常量，其他模块统一从此引用 |
| `data/` | 构建字符映射表、文本 → 索引序列、滑动窗口切分、DataLoader |
| `models/` | 定义 CharRNN：Embedding → LSTM×N → Dropout → Linear |
| `training/` | 训练循环、loss/困惑度计算、梯度裁剪、checkpoint 存取、日志记录 |
| `evaluation/` | 验证集困惑度评估、生成样本质量打分 |
| `inference/` | 给定 prompt 逐字采样，支持 temperature / top-k / top-p |
| `cli/` | `commands.py` 指令式子命令（train/generate/eval）；`menu.py` 交互式菜单 |
| `notebooks/` | Jupyter Notebook，用于数据探索和生成效果演示 |
| `scripts/` | 一次性工具脚本（如下载数据集） |
| `tests/` | 单元测试（模型 shape、数据加载、采样逻辑） |
| `outputs/` | checkpoints 存模型权重，logs 存训练日志，visualizations 存图表 |

### 数据流

`datasets/tangshi.txt` → `charVocab` (构建映射表) → `dataLoader` (索引 batch) → `charRNN.forward()` (logits) → `trainer` (loss → backward → step) → 每 N 个 epoch → `evaluator` (困惑度) + `generator` (采样诗句)

### 模型结构

`输入 (batchSize, seqLength)` → `Embedding [vocabSize × embeddingDim]` → `LSTM × 2 [hidden=256, dropout=0.3]` → `Linear [256 × vocabSize]` → `输出 (batchSize, seqLength, vocabSize)`

### CLI 两种模式

**指令模式** (`python main.py <command> [args]`)：

```
python main.py train          # 启动训练
python main.py generate       # 交互式文本生成
python main.py eval           # 评估模型
python main.py download       # 下载数据集
```

**交互菜单模式** (`python main.py`)：

```
========================================
  Char-RNN — 字符级文本生成
========================================
  1. 下载数据集
  2. 训练模型
  3. 生成文本
  4. 评估模型
  5. 查看训练日志
  0. 退出
========================================
请选择操作 [0-5]:
```

### 默认超参

| 参数 | 值 | 说明 |
|------|-----|------|
| `embeddingDim` | 128 | 字符嵌入维度 |
| `hiddenDim` | 256 | LSTM 隐藏层维度 |
| `numLayers` | 2 | LSTM 层数 |
| `dropout` | 0.3 | 层间 dropout |
| `seqLength` | 100 | 每个样本的字符数 |
| `batchSize` | 64 | 批次大小 |
| `learningRate` | 0.002 | 初始学习率 |
| `numEpochs` | 50 | 训练轮数 |
| `clipGrad` | 5.0 | 梯度裁剪阈值 |
| `temperature` | 0.8 | 生成时的随机性 |

## 三、实现路线

- [ ] **Phase 1：基础设施**
  - [ ] 创建项目目录结构
  - [ ] 实现 `config/defaults.py`（超参）和 `config/paths.py`（路径），所有路径自动创建
  - [ ] 实现 `scripts/downloadDataset.py`：下载全唐诗 JSON → 提取诗句 → 写入 `datasets/tangshi.txt`

- [ ] **Phase 2：数据模块**
  - [ ] 实现 `data/charVocab.py`：扫描文本构建 `char2idx` / `idx2char`，`encode()` / `decode()`，支持 `<UNK>`
  - [ ] 实现 `data/dataLoader.py`：索引序列 + 滑动窗口 (X, Y) + DataLoader
  - [ ] 编写测试验证 vocab 构建和 dataLoader 输出正确

- [ ] **Phase 3：模型模块**
  - [ ] 实现 `models/charRNN.py`：Embedding → LSTM×N → Dropout → Linear
  - [ ] `forward(x, hidden=None)` 返回 `(logits, hidden)`
  - [ ] 支持参数切换 `rnn_type`（RNN / LSTM / GRU）
  - [ ] 编写 `tests/testModel.py`：验证各 shape 正确

- [ ] **Phase 4：训练模块**
  - [ ] 实现 `training/trainer.py`：训练循环、交叉熵 loss、梯度裁剪、Adam 优化器
  - [ ] 每个 epoch 打印 `epoch / loss / perplexity / 耗时`
  - [ ] checkpoint 保存与加载（支持断点续训）
  - [ ] 训练日志写入 `outputs/logs/`

- [ ] **Phase 5：评估与推理模块**
  - [ ] 实现 `evaluation/evaluator.py`：验证集 perplexity 计算
  - [ ] 实现 `inference/generator.py`：temperature / top-k 采样、逐字续写
  - [ ] 训练中每 5 个 epoch 自动采样并写入 `outputs/logs/`

- [ ] **Phase 6：CLI 与主入口**
  - [ ] 实现 `cli/commands.py`：argparse 子命令（train / generate / eval / download）
  - [ ] 实现 `cli/menu.py`：交互式菜单，循环等待用户输入
  - [ ] 实现 `main.py`：无参数启动菜单模式，带参数走指令模式

- [ ] **Phase 7：可视化与 Notebook（可选）**
  - [ ] 绘制 loss 曲线，保存到 `outputs/visualizations/`
  - [ ] 对比不同 temperature 的生成效果
  - [ ] 编写 `notebooks/exploration.ipynb`：数据探索 + 生成效果演示

## 四、验收标准

- [ ] `python scripts/downloadDataset.py` 成功生成 `datasets/tangshi.txt`
- [ ] 前向传播 shape 正确：`(64, 100)` → `(64, 100, vocabSize)`
- [ ] 训练 50 epoch，loss 从 ~8.5 降至 ~3.0（perplexity ~20）
- [ ] 生成诗句每句 5 或 7 字，句式大致整齐
- [ ] `python main.py train` 一键启动训练
- [ ] `python main.py generate` 进入交互式文本生成
- [ ] `python main.py` 进入交互菜单，所有选项可正常使用
- [ ] 支持断点续训（中断后再次启动从 checkpoint 恢复）
